### ISSUE TITLE
Does this library totally ignore the accessibility?

### ISSUE THREAD

**Issue by changLiuUNSW** [reactstrap/reactstrap#655]
In the official Bootstrap 4  all the components written in JQuery are fully accessible. 
https://getbootstrap.com/docs/4.0/getting-started/accessibility/
But in this library, most components are not accessible at all. Does this library totally ignore the accessibility?
Actually most big companies like us Bank, we take accessibility very seriously when we choose a third part library.
We really hope the author of this library can take accessibility seriously like React-Bootstrap.
https://github.com/react-bootstrap/react-bootstrap/wiki#100-roadmap (See Accessibility section)

**Comment by TheSharpieOne** [comment]
Yes, this library is 508 compliant as much as bootstrap itself is. We have keyboard navigation/control of dropdowns and such, provide aria-* were applicable, and allow the user of the library to add their own attributes to all of the components (such as when tying a dropdownmenu to it's label via `aria-labelledby`. Since we do not generate IDs, the user would need to provide that).
No library is perfect, and as you can see other issues have been opened about a11y with specific components and those issues have been resolved. If you have specific conerns, please list them here or in individual issues so that they can be addressed.

**Comment by changLiuUNSW** [comment]
Take modal dialog for example (https://getbootstrap.com/docs/4.0/components/modal/):
1. Modal container has role = dialog
2. Modal content has role = document
3. Modal must be focused when it is opened
4. Tab and Shift+Tab do not move focus outside modal
5. When modal closes, focus returns to the element that had focus before the dialog was invoked.
6. Keyboard interaction Tab/Shift +Tab: move focus to the next/previous focusable element.
Similar to other components like menu, tooltip, accordion, tabs and etc. Actually all the accessibility implementation for those components can be found in Bootstrap 4 official website.

**Comment by TheSharpieOne** [comment]
1. Modal container has a default role of "dialog" but it is can overridden by providing you own role on `Modal`.
2. Yes, modal content has a role of "document" and that role cannot be overridden, is that the issue?
3. The Modal is focused when it is opened, you can even opt out of this by provided `autoFocus={false}` in case you want to focus on something within the modal (like an input) yourself.
4. Modal focus is an open issue #310, but even bootstrap's focus trap has a *major* a11y problem as it traps the focus too well, the keyboard user cannot tab to get to the browser's chrome. It completely traps the user on the page. This needs to be solved first. The only worse than not having an a11y feature is having a bad one.
5. This is related to 4, but probably could be it's ticket.
6. This is how the browser works. Use the right tabindex values or do not use them at all and the browser will tab to the next and previous focusable element automatically. You should also order your components element logically.

So 4 and 5 are the only issues, 4 is being tracked, but event bootstrap has a11y issues with it's implementation. 5 is related (and would probably be fixed at the same time), but can be it's own issue.

**Comment by changLiuUNSW** [comment]
Tooltips
https://getbootstrap.com/docs/4.0/components/tooltips/
1. missing ```role="tooltip"``` to the tooltip container
2. missing ```aria-describeby``` when it is open
Popover
https://getbootstrap.com/docs/4.0/components/popovers/
1. missing ```role="tooltip"``` to the popover container
2. missing ```aria-describeby``` when it is open
Tabs
https://getbootstrap.com/docs/4.0/components/navs/
1. missing ```role="tablist"``` to the ```ul```
2. missing ```role="tab"``` to the ```li```
Collapse
https://getbootstrap.com/docs/4.0/components/collapse/
1. missing ```aria-expanded``` and ```aria-controls``` for the toggle button
2. missing ```id``` for collapse container
Radio buttons and checkbox buttons
https://getbootstrap.com/docs/4.0/components/buttons/#button-plugin
1. semantically all radio buttons should be native radio input eg:
```
<label class="btn btn-secondary active">
    <input type="radio" name="options" id="option1" autocomplete="off" checked=""> Radio 1 (preselected)
  </label>
```
2. semantically all checkbox buttons should native checkbox eg:
```
<label class="btn btn-secondary">
    <input type="checkbox" checked="" autocomplete="off"> Checkbox 1 (pre-checked)
  </label>
```

**Comment by changLiuUNSW** [comment]
Here is the question I asked on Bootstrap 4 github page regarding their accessibility standard: 
https://github.com/twbs/bootstrap/issues/22406
You can see the reply from @patrickhlauke:
```
we strive towards making interactive controls/widgets expose all the correct roles/values/state properties (to satisfy, say, 4.1.2), to make things works with keyboard (2.1.1).
```
Actually most Jquery components they build are accessible for both keyboard and screen reader users and also meet WCAG 2.0 AA/AAA standard.
I think it is better to implement the same accessibility behavior with their JQuery components when we create React Bootstrap 4 components, because Accessibility is also a very important feature of Bootstrap 4.

**Comment by gergely-nagy** [comment]
I'd like to work on this.

**Comment by TheSharpieOne** [comment]
@gergely-nagy go for it

**Comment by joelwross** [comment]
Further issues: make sure you're not adding extraneous semantic HTML elements when using these components. For example, `<CardTitle>` has a default tag of `h4`, meaning that the "default" card examples have lots of extra headings. This interferes with navigation on Screen Readers. Instead, these components should have default tags that are not semantically loaded (e.g., `span`).

**Comment by gergely-nagy** [comment]
I didn't have much time in the last 2 months, but I will work on this issue again.

**Comment by mackbrowne** [comment]
I'm currently having some issues with the navbar.  The tab key / arrow keys for dropdowns work as expected.  However in all the examples the tab button will 'skip' all the links inside the dropdown when it's closed.  For me, it tabs through all the links inside the dropdown regardless if the dropdown is open.  

How do i fix this!?  I just want it to work like the examples.  I've updated all my packages to the latest and double checked that I have all the classnames the example has.  Not sure where to go from here....

@TheSharpieOne @changLiuUNSW any suggestions?  The boss is breathing down my neck for this one...

**Comment by TheSharpieOne** [comment]
@mackbrowne Are you able to post code or create a demo showcasing this (https://stackblitz.com/edit/reactstrap-v6?file=Example.js)

**Comment by mackbrowne** [comment]
@TheSharpieOne and anyone else, just reporting back.  I'm using animation to open/close the dropdown menu.  

I found [this](https://bootsnipp.com/snippets/vvMlB) snippet.  it ended up being the issue. 

```
display: block
```

 on the DropdownMenu component, for some reason breaks the tabbing on the navbar, no idea why but that was the issue.

**Comment by TheSharpieOne** [comment]
That snippet is for bootstrap v3, I may not be compatible with bootstrap v4 which is why you get that issue

**Comment by anevaude** [comment]
Came here after I noticed the Accordion (collapse) does not have Bootstraps aria-expanded controls.

https://getbootstrap.com/docs/4.0/components/collapse/#accessibility

Has there been a fix for this?

**Comment by TheSharpieOne** [comment]
@anevaude I opened #1627 to track that effort.

**Comment by anevaude** [comment]
> @anevaude I opened #1627 to track that effort.

Oh awesome, thanks! I'll follow that.
